### PR TITLE
Module1  react components

### PR DESCRIPTION
**Pull request by Shtuka1993**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
     "react-scripts": "5.0.1",
+    "sass": "^1.71.1",
```
Comment (esvyridov) [inline package.json]: Feel free to drop this dependency unless you would like to try it in the course. Nowadays these preprocessors aren't needed, especially if you support only modern browsers

File: src/common/Button/Button.tsx [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import * as React from 'react';
```
Comment (esvyridov) [inline src/common/Button/Button.tsx]: Starting from React 17, there is no need to import React to all "react" files. You can change field `jsx` from `react` to `react-jsx` and remove these imports.

File: src/common/Button/Button.tsx [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+interface buttonProps {
```
Comment (esvyridov) [inline src/common/Button/Button.tsx]:
```suggestion
interface ButtonProps {
```

File: src/common/Button/Button.tsx [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+	text: string;
+	onClick?(): void | React.MouseEvent<HTMLButtonElement>;
+}
+
+export default function Button(props: buttonProps): JSX.Element {
```
Comment (esvyridov) [inline src/common/Button/Button.tsx]: 👍 Another way to create a functional component (feel free to keep it as it is)

```ts
const Button: FC<ButtonProps> = ({ text, onClick }) => {
   ...
}
```

It's a bit shorter and it's nice that these is no need to type return type of the function. `FC` can be imported from `react` package

File: src/common/Button/Button.tsx [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+	const action = props.onClick !== undefined ? props.onClick : undefined;
```
Comment (esvyridov) [inline src/common/Button/Button.tsx]: I think this line is unnecessary, you can just pass `props.onClick` to the `onClick` prop because it accepts `undefined`

File: src/common/Button/Button.tsx [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import * as React from 'react';
+
+interface buttonProps {
+	text: string;
+	onClick?(): void | React.MouseEvent<HTMLButtonElement>;
```
Comment (esvyridov) [inline src/common/Button/Button.tsx]: This typing means that `onClick` doesn't accept any arguments and returns nothing or `React.MouseEvent<HTMLButtonElement>`, but this is not accurate because `onClick` accepts `React.MouseEvent<HTMLButtonElement>` and returns nothing, so I would suggest to change it to:

```suggestion
	onClick?: (e: MouseEvent<HTMLButtonElement>) => void;
```

File: src/App.scss [hunk 7]
```diff
@@ -0,0 +1,44 @@
+@use 'common/Button/Button.module.scss';
```
Comment (esvyridov) [inline src/App.scss]: I would like to suggest you to use CSS Modules in this course. They are quite popular nowadays and the idea behind it is simple.

So, instead of importing all .module.scss files in the common App.scss file, the modules can be imported in the `.tsx` files.

Let's take `Button.module.scss`:
1) (optional) rename it to `Button.module.css`
2) Import this file in `Button.tsx` as `import styles from './Button.module.css`
3) Class names should have camesCase format (`header-title` -> `headerTitle`). Snake case format can be used too, but in order to apply the class you need to use this syntax `styles['header-title']`.
4) Add a class to an element using `className` prop: `<button className={styles.button}>...</button>`

If you check name of the class in the `button` element in the browser, you see that the classname was transformed to a unique hash It's a good thing  that allows you to stop thinking if a class name is unique or not.

File: src/App.scss [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+@use 'common/Input/Input.module.scss';
+
+@use 'components/CourseInfo/CourseInfo.module.scss';
+@use 'components/Header/Header.module.scss';
+@use'components/Courses/components/SearchBar/SearchBar.module.scss';
+@use'components/Courses/components/CourseCard/CourseCard.module.scss';
+@use'components/Courses/Courses.module.scss';
+@use'components/EmptyCourseList/EmptyCourseList.module.scss';
+* {
```
Comment (esvyridov) [inline src/App.scss]: In most cases `*` is used only for changing `box-sizing` css property, so I would suggest to change it to


```suggestion
*,
*::after
*::before {
  box-sizing: border-box;
}

body {
  font-family: Open Sans;
  font-weight: 400;
  line-height: 1.5rem;
  font-size: 1rem;
  color: #333E48;
}
```

File: src/App.scss (hunk 7 above)
Comment (esvyridov) [reply src/App.scss]: Don't forget to rename `button` to `.button` in `Button.module.(s)css` in order to use it like this: `styles.button`

File: src/App.tsx [hunk 9]
```diff
@@ -1,26 +1,19 @@
...
-					rel='noopener noreferrer'
-				>
-					Learn React
-				</a>
-			</header>
-		</div>
+		<>
+			<Header />
+			<div className='content-block'>
+				<Courses data={mockedCoursesList} />
```
Comment (esvyridov) [inline src/App.tsx]: As we discussed today, you have two options:
1) Move state with courses to the App component from the Courses component and pass the value and the setter to Courses component as props

```suggestion
const [courses, setCourses] = useState(mockedCoursesList);
....
				<Courses courses={courses} updateCourses={setCourses}/>
```
2) Remove `data` prop,  import `mockedCoursesList` in the Courses and put it in the state there. So the same line
```
const [courses, setCourses] = useState(mockedCoursesList);
```
but in the Courses component

File: src/common/Input/Input.tsx [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import * as React from 'react';
+
+interface inputProps {
+	placeholder: string;
+	value: string;
+	name?: string;
+	label?: string;
+	onChange: (e: React.ChangeEvent) => void;
```
Comment (esvyridov) [inline src/common/Input/Input.tsx]: 👍 
```suggestion
	onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
```

File: src/common/Input/Input.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+export default function Input(props: inputProps): JSX.Element {
+	const { placeholder, value, name = '', label = '', onChange } = props;
+	const element: unknown =
```
Comment (esvyridov) [inline src/common/Input/Input.tsx]: In this case defining a type is unnecessary because Typescript is able to identify based on the ternary operator
```suggestion
	const element =
```

File: src/helpers/getCourses.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import { mockedCoursesList as data } from 'src/constants';
+
+export const getCourses = (query: string): object[] => {
+	const courses = data.map((item) => {
```
Comment (esvyridov) [inline src/helpers/getCourses.ts:4]: Only `.filter` can be used here to achieve the same result

```suggestion
  return data.filter((item) => {
    return item.id.includes(query) || item.title.includes(query)
    });
```

File: src/helpers/getCourses.ts [hunk 13]
```diff
@@ -0,0 +1,13 @@
+import { mockedCoursesList as data } from 'src/constants';
```
Comment (esvyridov) [inline src/helpers/getCourses.ts:1]: Feel free to keep it, but it's unnecessary

 ```suggestion
import { mockedCoursesList } from 'src/constants';
```

File: src/helpers/getCourseById.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import { mockedCoursesList as data } from 'src/constants';
+
+export const getCourseById = (id: string): unknown => {
```
Comment (esvyridov) [inline src/helpers/getCourseById.ts]: `.find` can be used here

```suggestion
export const getCourseById = (id: string): unknown => {
  return mockedCoursesList.find((course) => course.id === id)
}
```

File: src/components/Header/Header.module.scss [hunk 15]
```diff
@@ -0,0 +1,17 @@
+.header {
+    height: 80px;
+    padding: 12px 37px;
+    background-color: #fff;
+
+    display: flex;
+    justify-content: space-between;
+    align-items: center;
+
+    img {
```
Comment (esvyridov) [inline src/components/Header/Header.module.scss]: These two floats are needed because they have their parent is `flexbox` with `justify-content: space-between`

File: src/common/Input/Input.tsx [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+		label !== '' && name !== '' ? <label htmlFor={name}>{label}</label> : null;
+	return (
+		<>
+			{element}
```
Comment (esvyridov) [inline src/common/Input/Input.tsx:12]: You can also use the ternary operator in JSX

```suggestion
			{label && name ? <label ...>...</label> : null}
```

File: src/components/Courses/Courses.tsx [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+	data: object[];
+}
+
+export default function Courses(props: coursesProp) {
+	const { data } = props;
+
+	const [courses, setCourses] = useState(data);
+	const [showInfo, setShowInfo] = useState(false);
+	const [courseId, setCourseId] = useState('');
+	const components =
```
Comment (esvyridov) [inline src/components/Courses/Courses.tsx]: Let's try to put this JSX directly into the return statement, I think it should improve readability of this component

File: src/components/Courses/Courses.tsx [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+		);
+	const btn = <Button text={ADD_COURSE_BTN} />;
+	const empty = <EmptyCourseList />;
+	const list = <div>{components}</div>;
+	const view = <>{courses.length !== 0 ? list : empty}</>;
+	const component = showInfo ? (
+		<CourseInfo
+			course={getCourseById(courseId)}
+			setShowInfo={setShowInfo}
+			setCourses={setCourses}
```
Comment (esvyridov) [inline src/components/Courses/Courses.tsx]: Can we avoid passing `setCourses` into the `CourseInfo` component somehow? For me they seem unrelated to each other

File: src/components/Courses/Courses.tsx [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+			<></>
+		);
+	const btn = <Button text={ADD_COURSE_BTN} />;
+	const empty = <EmptyCourseList />;
+	const list = <div>{components}</div>;
+	const view = <>{courses.length !== 0 ? list : empty}</>;
+	const component = showInfo ? (
+		<CourseInfo
+			course={getCourseById(courseId)}
+			setShowInfo={setShowInfo}
```
Comment (esvyridov) [inline src/components/Courses/Courses.tsx]: Let's pass `onBackClick` or `onBackButton` click prop to `CourseInfo` and call it there when Back button is clicked.

```suggestion
			onBackClick={() => setShowInfo(false)}
```

File: src/components/Courses/Courses.tsx [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+	const [courses, setCourses] = useState(data);
+	const [showInfo, setShowInfo] = useState(false);
+	const [courseId, setCourseId] = useState('');
+	const components =
+		courses.length !== 0 ? (
+			courses.map((course, id) => {
+				return (
+					<CourseCard
+						course={course}
+						setShowInfo={setShowInfo}
```
Comment (esvyridov) [inline src/components/Courses/Courses.tsx]: This two props can be change to only one

```suggestion
						onShowCourseClick={(id) => {
						  setShowInfo(true);
						  setCourseId(id);
						 }
```

But you can also remove `showInfo` state and show CourseInfo only if `courseId` is defined

File: src/helpers/getCourseDuration.ts [hunk 21]
```diff
@@ -0,0 +1,7 @@
+export const getCourseDuration = (minutes: number): string => {
+	const hourDuration = 60;
+	const hours = Math.floor(minutes / hourDuration);
+	minutes = minutes % hourDuration;
```
Comment (esvyridov) [inline src/helpers/getCourseDuration.ts]: Hmm, let's create a new variable here, changing function arguments can be dangerous, especially if it's an object

File: src/helpers/formatAuthorsData.ts [hunk 22]
```diff
@@ -0,0 +1,9 @@
+import { mockedAuthorsList } from 'src/constants';
+
+export const formatAuthorsData = (ids: string[]): string => {
+	const authors = mockedAuthorsList.map((author) => {
```
Comment (esvyridov) [inline src/helpers/formatAuthorsData.ts]: You can find a bug on the home page in the Javascript course - `Authors:Vasiliy Dobkin, Nicolas Kim, ,`. This empty strings between two commas in the end are created because `.join` joins all variables of the array (including undefined).

I would use combination of filter, map and join here

```suggestion
	return mockedAuthorsList
	  .filter((author) => {...})
	  .map((author) => {...})
	  .join(', ');
```

Could you finalize the suggestion please?

File: src/components/Courses/Courses.tsx (hunk 19 above)
Comment (Shtuka1993) [reply src/components/Courses/Courses.tsx]: Thanks. Good idea, I agree to use it but I would like to discuss hooks(useEffect) so for now I used useEffect here. I am not sure that this is the best idea. Please take a look on it.

File: src/components/Courses/Courses.tsx (hunk 20 above)
Comment (Shtuka1993) [reply src/components/Courses/Courses.tsx]: Could you please elaborate on this more?

Comment (esvyridov) [reply src/components/Courses/Courses.tsx]: Sure, here is some pseudo code:

```ts
const [courseId, setCourseId] = useState<string | undefined>();

if (!courseId) {
  return courses.map((course) => <CourseCard course={course} onShowCourseClick={setCourseId} />
}

return (
  <CourseInfo course={courses.find((course) => course.id === courseId)} />
)
```

My suggestion is to move logic for changing active `courseId` from `CourseCard` (line 18 `setCourseId(course.id);`) to `Courses` component. By doing that Courses will be able to react to whatever happens in the CourseCard component rather than share its own responsibilities to child components.

Comment (Shtuka1993) [reply src/components/Courses/Courses.tsx]: Thanks

Comment (Shtuka1993) [reply src/components/Courses/Courses.tsx]: I would like to discuss this workflow if you don't mind.

Comment (esvyridov) [reply src/components/Courses/Courses.tsx]: Sure, let's skip for now and discuss later. I'll review the PR tomorrow

**Review by esvyridov — APPROVED** [review]

File: src/App.tsx [hunk 23]
```diff
@@ -1,26 +1,22 @@
-import React from 'react';
-import logo from './logo.svg';
 import './App.css';
```
Comment (esvyridov) [inline src/App.tsx:1]: Let's convert it to CSS Module too in the next PR

File: src/common/Button/Button.tsx [hunk 24]
```diff
@@ -0,0 +1,13 @@
+import * as types from 'src/types';
```
Comment (esvyridov) [inline src/common/Button/Button.tsx:1]: For the next PR: A type for a component props is rarely used in other parts of an application but in the component file itself (and test file for this component), so it's ok if you put `ButtonProps` in this file, but keep global types like `Course` or `User` in `types.ts`

File: src/common/Button/Button.tsx [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+
+import { FC } from 'react';
+
+import styles from './Button.module.css';
+
+export const Button: FC<types.ButtonProps> = ({ text, onClick }) => {
+	return (
+		<button className={styles.button} onClick={onClick}>
```
Comment (esvyridov) [inline src/common/Button/Button.tsx:9]: 👍
